core/src/main/scala/dagr/core/execsystem/TaskManagerLike.scala

```diff
@@ -0,0 +1,177 @@
+  abstract class BaseInfoAndNode(info: TaskExecutionInfo, node: BaseGraphNode)
+}
+
+/** A generic template for task managers */
+trait TaskManagerLike {
+
+  /** Gets the task associated with the identifier, if any
+    *
+    * @param id the task id
+    * @return the task associate with the id if found, None otherwise
+    */
+  def taskFor(id: BigInt): Option[Task]
+
+  /** Get the task's execution information if it exists.
+    *
+    * @param id the task identifier.
+    * @return the task execution information if the task is managed, None otherwise
+    */
+  def taskExecutionInfoFor(id: BigInt): Option[TaskExecutionInfo]
+
+  /** Get the task's execution information if it exists.
+    *
+    * @param task the task.
+    * @return the task execution information if the task is managed, None otherwise
+    */
+  def taskExecutionInfoFor(task: Task): Option[TaskExecutionInfo]
+
+  /** Get the task's associated [[TaskStatus]].
+    *
+    * @param task the task.
+    * @return the task status if the task is managed, None otherwise.
+    */
+  def taskStatusFor(task: Task): Option[TaskStatus.Value]
+
+
+  /** Get the task's associated [[TaskStatus]].
+    *
+    * @param id the task identifier.
+    * @return the task status if the task is managed, None otherwise.
+    */
+  def taskStatusFor(id: BigInt): Option[TaskStatus.Value]
+
+
+  /** Get the task identifier for the given task.
+    *
+    * @param task the task.
+    * @return the task identifier, None if the task is not being managed.
+    */
+  def taskIdFor(task: Task): Option[BigInt] = task.id
+
+  /** Get the task identifiers for all tracked tasks.
+    *
+    * @return the task ids
+    */
+  def taskIdsFor(): Iterable[BigInt]
+
+  /** Gets the task for the given id.  If there is no task being tracked
+    * with that id, throws a [[NoSuchElementException]].
+    *
+    * @param id the id to lookup.
+    * @return the associated task.
+    */
+  def apply(id: BigInt): Task
+
+  /** Gets the task execution info for the given task.  If the task is not
+    *  being tracked, throws a [[NoSuchElementException]].
+    *
+    * @param task the task to lookup.
+    * @return the associated task execution info.
+    */
+  def apply(task: Task): BaseInfoAndNode
+
+  /** Get the bi-directional map between managed tasks and their associated task execution information.
+    *
+    * @return the task and task execution information bi-directional map.
+    */
+  def taskToInfoBiMapFor: BiMap[Task, TaskExecutionInfo]
+
+  /** Checks if we have failed tasks.
+    *
+    * @return true if we have failed tasks, false otherwise.
+    */
+  def hasFailedTasks: Boolean
+
+  /** Adds a task to be managed
+    *
+    * @param task the given task.
+    * @return the task identifier.
+    */
+  def addTask(task: Task): BigInt
+
+  /** Adds tasks to be managed
+    *
+    * @param tasks the given tasks.
+    * @return the task identifiers.
+    */
+  def addTasks(tasks: Task*): List[BigInt]
```

Comment:
I'll also make it a `Seq`, since I want to return the ids in the same order, and is a bit more general than `List`.